### DISCUSSION TITLE
Inquiry about Embedding Extraction

### DISCUSSION THREAD

**Discussion by PhilipAmadasun** [pyannote/pyannote-audio#1663] [Q&A]
I have two questions about extracting embeddings
* First:
    from scipy.spatial.distance import cdist distance = cdist(embedding1, embedding2, metric="cosine")[0,0]
    Gives an error, I was following the tutorial [here](https://huggingface.co/pyannote/embedding)
    I have to do this instead:
    distance = cdist(np.expand_dims(embedding1,axis=0), np.expand_dims(embedding2, axis=0), metric="cosine")[0,0]
    To get results. And now I'm confused if I'm even supposed to do this.

 * Second:
    Do you get the most accurate "Essence" of someone's voice via embedding by:
        - Extracting embedding from longer audio clips of the persons talking?
        - Gathering (let's say) 30 millisecond chunks of audio, extracting the embeddings from each chunk, then getting an average embedding from them?
        Which one of these is the way to go?

**Comment by fungus75** [comment]
First: 
I guess they use nowadays speechbrain for identification of speakers: [Link](https://github.com/pyannote/pyannote-audio/blob/6580e6c8d6e4431778438ce1e7d591aae8688f58/pyannote/audio/pipelines/speaker_diarization.py#L119)

Therefore the similarity-Identification might be done somehow like [that](https://github.com/speechbrain/speechbrain/blob/a61a0dd7659b9ffcc7e2bf55d12af51e2d4b8ab0/speechbrain/inference/speaker.py#L51)/[that](https://github.com/speechbrain/speechbrain/blob/a61a0dd7659b9ffcc7e2bf55d12af51e2d4b8ab0/speechbrain/inference/speaker.py#L90-L91).

I tried out the following:
```
similarity = torch.nn.CosineSimilarity(dim=-1, eps=1e-6)
sim = similarity(torch.from_numpy(embedding1), torch.from_numpy(embedding2))
```

But for me the question is, how big this "sim" must be to be identical.
According their [code ](https://github.com/speechbrain/speechbrain/blob/a61a0dd7659b9ffcc7e2bf55d12af51e2d4b8ab0/speechbrain/inference/speaker.py#L54) 0.25 is a good value? For me it is way too low.

**Reply by PhilipAmadasun** [reply]
@fungus75 The way I previously described still works when it's people whose voice embeddings I've already saved. I don't have a good threshold to pick put unknown voices. I'll try this way you've mentioned.

**Reply by PhilipAmadasun** [reply]
@fungus75 Yeah after some brief testing I'm not sure their way can work for me. Have you tried the way I mentioned previously and compared the two methods? I'm curious. I'm kinda working in briefs waves due to illness currently, would appreciate more input on this as I can't carry out this task currently. Maybe it it will also help you as well hopefully.
